jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ResourceService.java

```diff
@@ -469,66 +466,130 @@ protected void sendWelcome(HttpContent content, String pathInContext, boolean en
 
-    public enum WelcomeActionType
+    /**
+     * <p>How welcome targets should be processed.</p>
+     */
+    public enum WelcomeMode
     {
+        /**
+         * The welcome target is used as the location for a redirect response,
+         * sent by {@link #redirectWelcome(Request, Response, Callback, String)}.
+         */
         REDIRECT,
-        SERVE
+        /**
+         * The welcome target is served by
+         * {@link #serveWelcome(Request, Response, Callback, String)}.
+         */
+        SERVE,
+        /**
+         * The welcome target is re-handled by
+         * {@link #rehandleWelcome(Request, Response, Callback, String)}.
+         */
+        REHANDLE
     }
 
     /**
-     * Behavior for a potential welcome action
-     * as determined by {@link ResourceService#processWelcome(Request, Response)}
-     *
-     * <p>
-     * For {@link WelcomeActionType#REDIRECT} this is the resulting `Location` response header.
-     * For {@link WelcomeActionType#SERVE} this is the resulting path to for welcome serve, note that
-     * this is just a path, and can point to a real file, or a dynamic handler for
-     * welcome processing (such as Jetty core Handler, or EE Servlet), it's up
-     * to the implementation of {@link ResourceService#welcome(Request, Response, Callback)}
-     * to handle the various action types.
-     * </p>
+     * <p>A welcome target paired with how to process it.</p>
      *
-     * @param type the type of action
-     * @param target The target URI path of the action.
+     * @param target the welcome target
+     * @param mode the welcome mode
      */
-    public record WelcomeAction(WelcomeActionType type, String target) {}
+    public record WelcomeAction(String target, WelcomeMode mode)
+    {
+    }
 
-    private boolean welcome(Request request, Response response, Callback callback) throws IOException
+    private boolean welcome(Request request, Response response, Callback callback) throws Exception
     {
-        WelcomeAction welcomeAction = processWelcome(request, response);
+        WelcomeAction welcomeAction = processWelcome(request);
         if (LOG.isDebugEnabled())
-        {
-            LOG.debug("welcome(req={}, resp={}, callback={}) welcomeAction={}",
-                request, response, callback, welcomeAction);
-        }
+            LOG.debug("welcome(req={}, rsp={}, cbk={}) welcomeAction={}", request, response, callback, welcomeAction);
+
         if (welcomeAction == null)
             return false;
 
-        handleWelcomeAction(request, response, callback, welcomeAction);
-        return true;
+        return handleWelcomeAction(request, response, callback, welcomeAction);
     }
 
-    protected void handleWelcomeAction(Request request, Response response, Callback callback, WelcomeAction welcomeAction) throws IOException
+    protected boolean handleWelcomeAction(Request request, Response response, Callback callback, WelcomeAction welcomeAction) throws Exception
```

Comment:
I think this should be a void method.  A welcome file has been found and there is no option but to serve it somehow.  If the configuration of the server does not allow it to be served, then it is a 500.